### PR TITLE
Use the correct `context` when re-using a cached `NodePath`

### LINKED ISSUE #12329: 7.12 BUG: Cannot read property 'hasOwnBinding' of null; class property; initializer; SystemJS
Repo: https://github.com/shrinktofit/babel-bug

## Bug Report

- [ ] I would like to work on a fix!

**Current behavior**

```
(node:2504) UnhandledPromiseRejectionWarning: TypeError: X:\Temp\babelbug\light.ts: Cannot read property 'hasOwnBinding' of null
    at Object.ReferencedIdentifier (X:\Temp\babelbug\node_modules\@babel\helper-create-class-features-plugin\lib\misc.js:35:20)
    at Object.newFn (X:\Temp\babelbug\node_modules\@babel\traverse\lib\visitors.js:216:17)
    at NodePath._call (X:\Temp\babelbug\node_modules\@babel\traverse\lib\path\context.js:55:20)
    at NodePath.call (X:\Temp\babelbug\node_modules\@babel\traverse\lib\path\context.js:42:17)
    at NodePath.visit (X:\Temp\babelbug\node_modules\@babel\traverse\lib\path\context.js:92:31)
    at TraversalContext.visitQueue (X:\Temp\babelbug\node_modules\@babel\traverse\lib\context.js:112:16)
    at TraversalContext.visitSingle (X:\Temp\babelbug\node_modules\@babel\traverse\lib\context.js:84:19)
    at TraversalContext.visit (X:\Temp\babelbug\node_modules\@babel\traverse\lib\context.js:140:19)
    at Function.traverse.node (X:\Temp\babelbug\node_modules\@babel\traverse\lib\index.js:82:17)
    at NodePath.visit (X:\Temp\babelbug\node_modules\@babel\traverse\lib\path\context.js:99:18)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:2504) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:2504) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

- [REPL](babeljs.io/repl), [Codesandbox](https://codesandbox.io/s/babel-repl-custom-plugin-7s08o?file=/src/index.js), or GitHub Repo helps!

**Input Code**

```ts
import { Vec3 } from '../../math';

export class Light {
    protected _color: Vec3 = new Vec3(1, 1, 1);
}

```

**Expected behavior**

No error.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

```js
const babel = require('@babel/core');
const presetEnv = require('@babel/preset-env');
const fs = require('fs-extra');
(async () => {
    const result = await babel.transformFileAsync(String.raw`light.ts`, {
        presets: [
            [presetEnv, {
                modules: 'systemjs',
            }],
            [require('@babel/preset-typescript')],
        ],
        plugins: [
            [require('@babel/plugin-proposal-class-properties')],
        ],
    });
    console.log(result.code);
})();

```

**Environment**

```
  System:
    OS: Windows 10 10.0.19041
  Binaries:
    Node: 14.12.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.17.3 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.14.7 - C:\Program Files\nodejs\npm.CMD
  npmPackages:
    @babel/core: 7.12.3 => 7.12.3
    @babel/plugin-proposal-class-properties: 7.12.1 => 7.12.1
    @babel/preset-env: 7.12.1 => 7.12.1
    @babel/preset-typescript: 7.12.1 => 7.12.1
```
- Babel version(s): [e.g. v6.0.0, v7.0.0-beta.34]
- Node/npm version: [e.g. Node 8/npm 5]
- OS: [e.g. OSX 10.13.4, Windows 10]
- Monorepo: [e.g. yes/no/Lerna]
- How you are using Babel: [e.g. `cli`, `register`, `loader`]

**Possible Solution**

**Additional context**

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12329
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  | ?
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

When re-using a cached path, by default we didn't update the context/opts. This means that, if I traverse a node with `{ noScope: true }` and then re-traverse it with `{ noScope: false }`, the second one will still have `noScope: true` because it re-uses the cached path with the original context.

The error in #12329 happens because we introduced `noScope: true` [in a traversal](https://github.com/babel/babel/blob/32279147e6a69411035dd6c43dc819d668c74466/packages/babel-helper-module-transforms/src/rewrite-this.js#L7) used by the SystemJS plugin in `Program:enter`, and that leaked in to the class properties plugin, causing `constructor.scope` to be undefined [in the `helper-class-features` package](https://github.com/babel/babel/blob/32279147e6a69411035dd6c43dc819d668c74466/packages/babel-helper-create-class-features-plugin/src/misc.js#L69).

This is _technically_ a breaking change (we are changing the default value of a parameter in the public API), but I couldn't find anyone online using that parameter and I'm not sure why anyway would _not_ wont it to use the correct context.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12331"></a>

### CODE REVIEW

File: packages/babel-preset-env/test/fixtures/sanity/block-scoping-for-of/output.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(
...
+function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e2) { throw _e2; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e3) { didErr = true; err = _e3; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
```
Comment (nicolo-ribaudo) [inline packages/babel-preset-env/test/fixtures/sanity/block-scoping-for-of/output.js:9]: I'm not sure why the tmp var name is changing here.

File: packages/babel-plugin-transform-parameters/test/fixtures/regression/6057-expanded/output.js [hunk 2]
```diff
@@ -9,8 +11,6 @@ var _args = _interopRequireDefault(require("utils/url/args"));
 
 function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
 
-function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-parameters/test/fixtures/regression/6057-expanded/output.js:12]: Ironically, this test was about another `context` caching bug (https://github.com/babel/babel/issues/6057), but this change has nothing to do with that bug.

File: scripts/integration-tests/e2e-babel-old-version.sh [hunk 3]
```diff
@@ -43,6 +43,10 @@ node "$PWD"/scripts/integration-tests/utils/bump-babel-dependencies.js
   fs.writeFileSync('./package.json', JSON.stringify(pkg, null, 2));
 "
 
+# https://github.com/babel/babel/pull/12331 - This test is fixed in @babel/traverse 7.12.7,
+# so it will fail with older versions. We can disable it.
+(cd packages/babel-plugin-transform-modules-systemjs/test/fixtures/regression/; mv issue-12329 .issue-12329)
```
Comment (JLHwung) [inline scripts/integration-tests/e2e-babel-old-version.sh:48]: What about supporting `minBabelVersion` in fixture runner 😄

Comment (nicolo-ribaudo) [reply scripts/integration-tests/e2e-babel-old-version.sh:48]: Only if this problem comes up more than once :stuck_out_tongue:

**Review by JLHwung — APPROVED** [review]
The `context` parameter was implemented in https://github.com/babel/babel/commit/4c0b8599f183d979c7fae912ffa16e0959858de1. I believe it is an unwanted side effect that by default `NodePath#get` does not reset correct context.

**Review by existentialism — APPROVED** [review]

**Comment by salvoravida** [comment]
image



after this pr, my  project crash ...

**Comment by salvoravida** [comment]
"This is technically a breaking change (we are changing the default value of a parameter in the public API), but I couldn't find anyone online using that parameter and I'm not sure why anyway would not wont it to use the correct context."

What do you think about semver?

You are breaking many production CI build .... bah...

**Comment by ycjcl868** [comment]
image

**Comment by nicolo-ribaudo** [comment]
@ycjcl868 I'm not sure how it is related to this PR, could you open a new issue? :pray:

**Comment by ycjcl868** [comment]
When I reset this line code, the build is passed.

> I think this line code would make `import()` syntax convert into `require()` syntax before the other babel plugin.

image

**Comment by nicolo-ribaudo** [comment]
Thanks, investigating.

**Comment by JLHwung** [comment]
@ycjcl868 It's after midnight in Hangzhou, have a good night. We are investigating, a patch will be released today.

**Pull request merged**